rslaser/rspulse/pulse.py

```diff
@@ -26,24 +26,50 @@
     'propLen', 'pulseE', 'poltype', 'sampFact',
     'mx', 'my', 'numsig']
 
-
-def check_type_and_fields(input_obj, req_fields, exception, class_name):
-    if type(input_obj) != PKDict:
-        raise exception(f'invalid inputs: parameters to {class_name} class must be of type PKDict')
-    for p in req_fields:
-        if p not in input_obj:
-            raise exception(f'invalid inputs: missing required field {p}')
-
-
-def _validate_input(input_params):
-    check_type_and_fields(input_params, _REQUIRED_LASER_PULSE_INPUTS, InvalidLaserPulseInputError, 'LaserPulse')
-    check_type_and_fields(input_params.slice_params, _REQUIRED_LASER_PULSE_SLICE_INPUTS, InvalidLaserPulseInputError, 'LaserPulseSlice')
-
-
-def _validate_input_slice(input_params, index):
-    if type(index) != int:
-        raise InvalidLaserPulseInputError(f'invalid inputs: LaserPulseSlice index must be of type int')
-    _validate_input(input_params)
+_LASER_PULSE_SLICE_DEFAULTS = PKDict(
+        sigrW=0.000186,
+        propLen=15,
+        pulseE=0.001,
+        poltype=1,
+        sampFact=1,
+        numsig=3.,
+        mx=0,
+        my=0
+    )
+_LASER_PULSE_DEFAULTS = PKDict(
+        phE=1.55,
+        nslice=3,
+        chirp=0,
+        w0=.1,
+        a0=.01,
+        dw0x=0.0,
+        dw0y=0.0,
+        z_waist=0,
+        dzwx=0.0,
+        dzwy=0.0,
+        tau_fwhm=0.1 / const.c / math.sqrt(2.),
+        z_center=0,
+        x_shift = 0.,
+        y_shift=0.,
+        d_to_w=0,
+        slice_params=_LASER_PULSE_SLICE_DEFAULTS,
+)
+
+
+def check_fields(input_obj, req_fields, exception, class_name):
+    for p in input_obj:
+        if p not in req_fields:
```

Comment:
I don't know how this all is going to be used so take this with a grain of salt. But, from my reading of the code I don't see an advantage to this test. What difference does it make if someone supplies a param and it isn't one we want to use? Generally, we try to be flexible in what we accept and strict in what we produce. If someone supplies an argument we don't know what to do with we just ignore it. I think strict type checking like this is a poor match for a dynamic language like python.

Reply:
Moving my previous comment to the correct location:

Regarding the type checking, these are the problems we are trying to prevent:

The user specifies a physical parameter with a typo in the name. In this case, the user input will be ignored, while the class constructor uses the default value. As a result, the user will be getting results inconsistent with what they think they specified.

A confused user specifies a physical quantity like laser wavelength. However, the constructor only considers photon energy as input, so the default photon energy is used, while the user-specified wavelength is ignored.

Basically, we're recovering the level of argument validation that one expects from a constructor with a long list of named arguments.

Reply:
Generally you are right @e-carlin, but if for example a user inputed photonic energy as photon_energy rather than phE and was getting weird results it would be kind of like a silent failure we are trying to avoid

Reply:
Thanks for that info. I think the check makes sense for the behavior you're after.